reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -121,18 +174,43 @@ public LoadBalancer(Publisher<List<ReactiveSocketFactory<SocketAddress>>> factor
     }
 
     private synchronized void addSockets(int numberOfNewSocket) {
-        activeFactories.entrySet()
-            .stream()
-            // available factories that don't map to an already established socket
-            .filter(e -> !activeSockets.containsKey(e.getKey()))
-            .map(e -> e.getValue())
-            .filter(factory -> factory.availability() > 0.0)
-            .sorted((a, b) -> -Double.compare(a.availability(), b.availability()))
-            .limit(numberOfNewSocket)
-            .forEach(factory -> {
-                pendingSockets += 1;
-                factory.apply().subscribe(new SocketAdder(factory.remote()));
-            });
+        int n = Math.min(activeFactories.size(), numberOfNewSocket);
+        int maxEffort = 1 + activeFactories.size()/2;
+        int effort = 0;
+        while (n > 0 && effort < maxEffort) {
+            ReactiveSocketFactory<?> factory0 = activeFactories.poll();
+            ReactiveSocketFactory<?> factory1 = activeFactories.poll();
+            if (factory0 == null) {
+                // queue is empty
+                break;
+            }
+            if (factory1 == null) {
+                // factory0 was the last item in the queue
+                if (factory0.availability() > 0.0) {
+                    pendingSockets++;
+                    factory0.apply().subscribe(new SocketAdder(factory0));
+                } else {
+                    activeFactories.addLast(factory0);
+                }
+                break;
+            }
+
+            if (factory0.availability() == 0.0 && factory1.availability() == 0.0) {
+                activeFactories.addLast(factory0);
+                activeFactories.addLast(factory1);
+                effort++;
+            } else if (factory0.availability() < factory1.availability()) {
+                n--;
+                pendingSockets++;
+                activeFactories.addLast(factory0);
```

Comment:
It's hard to reason about this load balancer because without the random element it seems affected by its environment.  

e.g. it seems like FactoriesRefresher will add new elements to the tail of the list, so something like a cluster roll will likely result in waves of traffic to the same shards at roughly the same time.